Replace local footer in existing sheared overlays (e.g. mod select & first-run setup) with `ScreenFooter`

frenzibyte opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/28682

Preview:

https://github.com/ppy/osu/assets/22781491/a5a5f034-ab01-4e95-913c-fdd8c5f8d320

https://github.com/ppy/osu/assets/22781491/0c0a5cd9-56be-4025-a127-5a617cd2950b

frenzibyte committed:
Migrate mod select overlay footer content

frenzibyte committed:
Migrate first-run setup overlay footer content

frenzibyte committed:
Refactor game-wide layout order of footer to fix depth issues with overlays and improve UX

With this new order, the logo can be easily moved to display in front of the footer in `SongSelectV2` without breaking experience when footer-based overlays are present. Such overlays (i.e. mod select overlay) will also be dimmed alongside the current screen when a game-wide overlay is open (e.g. settings).

frenzibyte committed:
Bind game-wide mods bindable to mod select overlay in new song select screen

frenzibyte committed:
Remove local footer from `ShearedOverlayContainer`

frenzibyte committed:
Fix first-run setup buttons reset after reopening from dismiss

frenzibyte committed:
Merge branch 'master' into footer-v2-integration

peppy committed:
Merge branch 'master' into footer-v2-integration

peppy commented on osu.Game/OsuGame.cs:
```diff
@@ -290,6 +297,8 @@ public void CloseAllOverlays(bool hideToolbar = true)
             if (hideToolbar) Toolbar.Hide();
         }
 
+        public void ChangeLogoDepth(bool inFrontOfFooter) => ScreenContainer.ChangeChildDepth(logoContainer, inFrontOfFooter ? float.MinValue : 0);
```
I dunno about this. I'd rather just see it using `Parent` access similar to other usages in `LogoTrackingContainer`...

frenzibyte replied:
I don't follow what you mean by "`Parent` access".

peppy replied:
`Parent.ChangeChildDepth` rather than `OsuGame` method.

frenzibyte replied:
Well...the drawable which its depth is being changed here is the parent of `OsuLogo`, meaning that this method is the equivalent of `logo.Parent.Parent.ChangeChildDepth(logo.Parent, float.MinValue)`, and not to mention that I can't access `ChangeChildDepth` directly without direct-casting to `Container` first. That's why I settled for an `OsuGame` method to ease my head.

peppy replied:
Right. It's still not great. See if you can find a nicer way of doing this, else I'll take a look during next review pass.

frenzibyte replied:
I have already spent hours before making this PR trying to figure out how to fix this issue of the logo depth being a nuisance in the user's experience, but I'll try once more and see if I have a clever idea this time.

peppy commented on osu.Game/Overlays/FirstRunSetup/ScreenUIScale.cs:
```diff
@@ -166,7 +168,8 @@ private void load(AudioManager audio, TextureStore textures, RulesetStore rulese
                                 Position = new Vector2(0.5f),
-                            })
+                            }),
+                            (typeof(ScreenFooter), footer = new ScreenFooter()),
```
Maybe comment on these that they are temporary until a global footer is present?

frenzibyte replied:
This one is a special case, we want to detach the `SongSelect` nested in this screen away from the global footer, because the global footer is already used to present the footer content for the first-run setup overlay. This is exactly like how the `OsuLogo` is cached here right above this line.

peppy replied:
right, makes sense in context.

peppy commented on osu.Game/Overlays/FirstRunSetupOverlay.cs:
```diff
@@ -190,26 +145,49 @@ protected override void LoadComplete()
+
+        private FirstRunSetupFooterContent? currentFooterContent;
+
+        public override Drawable CreateFooterContent()
```
I'd recommend changing the return type here to be more explicit:

```diff
diff --git a/osu.Game.Tests/Visual/UserInterface/TestSceneScreenFooter.cs b/osu.Game.Tests/Visual/UserInterface/TestSceneScreenFooter.cs
index de2026e538..a747a19684 100644
--- a/osu.Game.Tests/Visual/UserInterface/TestSceneScreenFooter.cs
+++ b/osu.Game.Tests/Visual/UserInterface/TestSceneScreenFooter.cs
@@ -210,7 +210,7 @@ public override bool OnBackButton()
                 return false;
             }
 
-            public override Drawable CreateFooterContent() => new TestFooterContent();
+            public override VisibilityContainer? CreateFooterContent() => new TestFooterContent();
 
             public partial class TestFooterContent : VisibilityContainer
             {
diff --git a/osu.Game/Overlays/FirstRunSetupOverlay.cs b/osu.Game/Overlays/FirstRunSetupOverlay.cs
index 6412297663..fe836ba009 100644
--- a/osu.Game/Overlays/FirstRunSetupOverlay.cs
+++ b/osu.Game/Overlays/FirstRunSetupOverlay.cs
@@ -150,7 +150,7 @@ protected override void LoadComplete()
 
         private FirstRunSetupFooterContent? currentFooterContent;
 
-        public override Drawable CreateFooterContent()
+        public override VisibilityContainer? CreateFooterContent()
         {
             currentFooterContent = new FirstRunSetupFooterContent
             {
diff --git a/osu.Game/Overlays/Mods/ModSelectOverlay.cs b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
index bd04a1f6b3..8d11807875 100644
--- a/osu.Game/Overlays/Mods/ModSelectOverlay.cs
+++ b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
@@ -256,7 +256,7 @@ protected override void LoadComplete()
 
         private ModSelectFooterContent? currentFooterContent;
 
-        public override Drawable CreateFooterContent() => currentFooterContent = new ModSelectFooterContent(this)
+        public override VisibilityContainer? CreateFooterContent() => currentFooterContent = new ModSelectFooterContent(this)
         {
             Beatmap = { BindTarget = Beatmap },
             ActiveMods = { BindTarget = ActiveMods },
diff --git a/osu.Game/Overlays/Mods/ShearedOverlayContainer.cs b/osu.Game/Overlays/Mods/ShearedOverlayContainer.cs
index 9e5a336c17..9ea98c1ae4 100644
--- a/osu.Game/Overlays/Mods/ShearedOverlayContainer.cs
+++ b/osu.Game/Overlays/Mods/ShearedOverlayContainer.cs
@@ -104,7 +104,7 @@ private void load()
         /// <summary>
         /// Creates content to be displayed on the game-wide footer.
         /// </summary>
-        public virtual Drawable CreateFooterContent() => Empty();
+        public virtual VisibilityContainer? CreateFooterContent() => null;
 
         /// <summary>
         /// Invoked when the back button in the footer is pressed.
diff --git a/osu.Game/Screens/Footer/ScreenFooter.cs b/osu.Game/Screens/Footer/ScreenFooter.cs
index a841f2a50b..4464b9d7da 100644
--- a/osu.Game/Screens/Footer/ScreenFooter.cs
+++ b/osu.Game/Screens/Footer/ScreenFooter.cs
@@ -218,7 +218,7 @@ public void SetActiveOverlayContainer(ShearedOverlayContainer overlay)
 
             updateColourScheme(overlay.ColourProvider.ColourScheme);
 
-            var content = overlay.CreateFooterContent();
+            var content = overlay.CreateFooterContent() ?? Empty();
 
             Add(contentContainer = new Container
             {
diff --git a/osu.Game/Screens/OnlinePlay/FreeModSelectOverlay.cs b/osu.Game/Screens/OnlinePlay/FreeModSelectOverlay.cs
index 0ed45161f2..d3eaa91447 100644
--- a/osu.Game/Screens/OnlinePlay/FreeModSelectOverlay.cs
+++ b/osu.Game/Screens/OnlinePlay/FreeModSelectOverlay.cs
@@ -6,6 +6,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
 using osu.Game.Graphics.UserInterface;
 using osu.Game.Overlays.Mods;
 using osu.Game.Rulesets.Mods;
@@ -34,7 +35,7 @@ public FreeModSelectOverlay()
 
         protected override ModColumn CreateModColumn(ModType modType) => new ModColumn(modType, true);
 
-        public override Drawable CreateFooterContent() => currentFooterContent = new FreeModSelectFooterContent(this)
+        public override VisibilityContainer? CreateFooterContent() => currentFooterContent = new FreeModSelectFooterContent(this)
         {
             Beatmap = { BindTarget = Beatmap },
             ActiveMods = { BindTarget = ActiveMods },

```

peppy reviewed (commented):
Tests need fixing too

frenzibyte committed:
Ask for `VisibilityContainer` explicitly

frenzibyte committed:
Add failing test case for edge case

frenzibyte committed:
Fix loading (but not showing) a sheared overlay hiding displayed footer content

Identified by tests. See https://github.com/ppy/osu/actions/runs/9869382635/job/27253010485 & https://github.com/ppy/osu/actions/runs/9869382635/job/27253009622.

This change also prevents the initial `PopOut` call in overlays from calling `clearActiveOverlayContainer`, since it's not in the update thread and it's never meant to be called at that point anyway (it's supposed to be accompanied by a previous `PopIn` call adding the footer content).

frenzibyte commented:
I've applied a minor refactor on the footer content flow between `ShearedOverlayContainer` and `ScreenFooter` to prevent an overlay from incorrectly clearing footer content of another displayed overlay.

peppy commented:
@frenzibyte this is still failing tests. please run them locally.

frenzibyte commented:
Right, my apologies.

frenzibyte committed:
Fix `ModSelectOverlay` not hiding without a footer

frenzibyte committed:
Add screen footer to `ScreenTestScene`

frenzibyte committed:
Fix footer content not accessible by overlay when overriden by a subclass

frenzibyte committed:
Update `TestSceneFreeModSelectOverlay` to work again

peppy committed:
Move logo depth to a forward passing

peppy committed:
Move `logoContainer` local again

peppy commented:
Gonna get this in. It seems okay enough to move forward with.
